internal/decryption/output.go

```diff
@@ -21,39 +21,87 @@ import (
 	"go.mongodb.org/mongo-driver/bson"
 )
 
-type AuditLogOutput struct {
-	Warningf  func(lineNb int, format string, a ...interface{}) error
-	Error     func(lineNb int, err error) error
-	Errorf    func(lineNb int, format string, a ...interface{}) error
-	LogRecord func(lineNb int, logRecord interface{}) error
-}
-
-func buildOutput(out io.Writer) AuditLogOutput {
-	newLine := []byte{'\n'}
-	writeLine := func(value []byte) error {
-		if _, err := out.Write(value); err != nil {
-			return err
-		}
-		_, err := out.Write(newLine)
+type AuditLogOutput interface {
+	Warningf(lineNb int, format string, a ...interface{}) error
+	Error(lineNb int, err error) error
+	Errorf(lineNb int, format string, a ...interface{}) error
+	LogRecord(lineNb int, logRecord interface{}) error
+}
+
+type auditLogOutputImpl struct {
+	out io.Writer
+}
+
+type AuditLogErrorLevel string
+
+const (
+	AuditLogErrorLevelError   AuditLogErrorLevel = "decryptionLogError"
+	AuditLogErrorLevelWarning AuditLogErrorLevel = "decryptionLogWarning"
+)
+
+type AuditLogError struct {
+	Level AuditLogErrorLevel
+	Line  int
+	Err   error
+}
+
+func (e AuditLogError) MarshalBSON() ([]byte, error) {
+	doc := bson.D{
+		bson.E{Key: "atype", Value: string(e.Level)},
+		bson.E{Key: "line", Value: e.Line},
+		bson.E{Key: "error", Value: e.Err.Error()},
+	}
+	return bson.Marshal(doc)
+}
+
+func (e AuditLogError) Error() string {
+	return e.Err.Error()
+}
```

Comment:
[q] should we implement `func (e *AuditLogError) Is(target error) bool {`?

Reply:
I don't think so according to https://pkg.go.dev/errors#Is we should only do it if we want to treat other know errors (like `os.FileNotExists`) the same as our custom error. Which is not the case.